homeassistant/components/recorder/run_history.py

```diff
@@ -0,0 +1,144 @@
+"""Track recorder run history."""
+from __future__ import annotations
+
+import bisect
+from dataclasses import dataclass
+from datetime import datetime
+
+from sqlalchemy.orm.session import Session
+
+import homeassistant.util.dt as dt_util
+
+from .models import RecorderRuns, process_timestamp
+
+
+@dataclass
+class _RecorderRunsHistory:
+    """Bisectable history of RecorderRuns."""
+
+    run_timestamps: list[int]
+    runs_by_timestamp: dict[int, RecorderRuns]
+
+
+class RunHistory:
+    """Track recorder run history."""
+
+    def __init__(self) -> None:
+        """Track recorder run history."""
+        self._recording_start = dt_util.utcnow()
+        self._current_run_info: RecorderRuns | None = None
+        self._run_history = _RecorderRunsHistory([], {})
+
+    @property
+    def recording_start(self) -> datetime:
+        """Return the time the recorder started recording states.
+
+        This function is thread safe.
+        """
+        return self._recording_start
+
+    @property
+    def current(self) -> RecorderRuns:
+        """Get the current run.
+
+        This function is thread safe.
+        """
+        assert self._current_run_info is not None
+        return self._current_run_info
```

Comment:
The functions are sort of thread safe by themselves, but calling `recording_start`, and then `current` can give misleading results.

Reply:
I removed the claim that they are thread safe.